catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java

```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
+							if (checkedTags.size() >= MAX_NUMBER_OF_TAGS_CHECKED) {
+								((AlertDialog) dialog).getListView().setItemChecked(indexSelected, false);
+							} else {
+								checkedTags.add(availableTags[indexSelected]);
+							}
+						} else {
+							checkedTags.remove(availableTags[indexSelected]);
+						}
+					}
+				})
+				.setPositiveButton(getText(R.string.next), new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						project.setTags(checkedTags);
+						showProgressDialogAndUploadProject();
+					}
+				})
+				.setNegativeButton(getText(R.string.cancel), new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						Utils.invalidateLoginTokenIfUserRestricted(ProjectUploadActivity.this);
+						setShowProgressBar(false);
+						setNextButtonEnabled(true);
+					}
+				})
+				.setCancelable(false)
+				.show();
 	}
 
-	private void onNextButtonClick() {
-		String name = nameInputLayout.getEditText().getText().toString().trim();
+	private void showProgressDialogAndUploadProject() {
+		final String name = nameInputLayout.getEditText().getText().toString().trim();
 		String description = descriptionInputLayout.getEditText().getText().toString().trim();
 
-		String error = textChangedListener.validateInput(name);
-
-		if (error != null) {
-			nameInputLayout.setError(error);
-			invalidateOptionsMenu();
-			return;
+		if (!project.getName().equals(name)) {
+			ProjectRenameTask
+					.task(project.getName(), name);
+			ProjectLoadTask
+					.task(name, this);
+			project = ProjectManager.getInstance().getCurrentProject();
 		}
 
-		Project project = ProjectManager.getInstance().getCurrentProject();
+		project.setDescription(description);
+		project.setDeviceData(this);
+
+		File projectDir = new File(DEFAULT_ROOT_DIRECTORY, project.getName());
 
-		if (Utils.isDefaultProject(project, this)) {
-			nameInputLayout.setError(getString(R.string.error_upload_default_project));
+		final SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
+		String token = sharedPreferences.getString(Constants.TOKEN, Constants.NO_TOKEN);
+		String username = sharedPreferences.getString(Constants.USERNAME, Constants.NO_USERNAME);
+
+		Intent intent = new Intent(this, ProjectUploadService.class);
+
+		String[] sceneNames = project.getSceneNames().toArray(new String[0]);
+
+		final int notificationId = StatusBarNotificationManager.getInstance()
+				.createUploadNotification(this, name);
+
+		uploadProgressDialog = new AlertDialog.Builder(this)
+				.setTitle(getString(R.string.upload_project_dialog_title))
+				.setView(R.layout.dialog_upload_project_progress)
+				.setPositiveButton(R.string.progress_upload_dialog_show_program, null)
+				.setNegativeButton(R.string.done, new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int which) {
```

Comment:
here the variable "which" is again. down on 354 the name if the variable is "id". Would be clearer if they had the same name.

Reply:
![image](https://user-images.githubusercontent.com/6851559/53286328-2b5d2b80-376d-11e9-836b-d27f7448ac4f.png)

Reply:
Hi, sorry I don't get it. Are these two different DialogInterfaces? Or are the varnames mixed?
ProjectUploadActivity, 288, 326 -- varname = which
ProjectUploadActivity, 237, 244, 339 -- varname = id

Sorry for this pedantic stuff again. I really spent a lot of time trying to find / understand a difference.

Reply:
this is just android studio's auto-create. and TBH I don't usually check every auto-created varname with every name that was already in the code somewhere.